content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,354 @@
+```
+
+### Defining a Functional Type
+
+A functional type can be defined by using the syntax `(in) -> out`. Using the functions from the previous examples, the types of the functions are:
+
+```lua
+type add = (x: number, y: number) -> number
+type FindSource = (script: BaseScript, pattern: string) -> (string, number)
+```
+
+## Table Types
+
+Luau does not have a `table` type, instead, table types are defined using `{}` syntax. One way of defining tables is using the `{type}` syntax, which defines a list type.
+
+```lua
+local numbers: {number} = {1, 2, 3, 4, 5}
+local characterParts: {Instance} = LocalPlayer.Character:GetChildren()
+```
+
+Index types can be defined using `{[indexType] = valueType}`.
+
+```lua
+local frobulator: {[string]: any} = {
+	Foo = "bar",
+	Baz = 10
+}
+```
+
+Tables can also have explicit string indices defined in a type.
+
+```lua
+type Car = {
+	Speed: number,
+	Drive: (Car) -> ()
+}
+```
+
+## Variadics
+
+Here's a function that calculates the sum of an arbitrary amount of numbers.
+
+```lua
+local function addLotsOfNumbers(...)
+	local sum = 0
+
+	for _, v in {...} do
+		sum += v
+	end
+
+	return sum
+end
+```
+
+As expected, this can take any value, and the typechecker won't raise a warning if you give it an invalid type, such as a `string`.
+
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5)) -- 15
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- attempt to add string to number
+```
+
+Instead, it is possible to assign a type to the `...`, just like how other types are assigned.
+
+```lua
+local function addLotsOfNumbers(...: number)
+```
+
+And now, the second line raises a type error.
+
+```lua
+print(addLotsOfNumbers(1, 2, 3, 4, 5))
+print(addLotsOfNumbers(1, 2, "3", 4, 5)) -- TypeError: Type `string` could not be converted into `number`.
+```
+
+However, with a type definition, the following statement does not work.
+
+```lua
+type addLotsOfNumbers = (...: number) -> number -- Expected type, got ':'
+```
+
+Instead, use the syntax `...type` to define a variadic type.
+
+```lua
+type addLotsOfNumbers = (...number) -> number
+```
+
+## Unions and Intersections
+
+It is possible to define a type to be two or more types, using a union or intersection.
+
+```lua
+type NumberOrString = number | string
+type Object1 = {foo: string}
+type Object2 = {bar: number}
+type Object1and2 = Object1 & Object2 -- {foo: string} & {bar: number}
+```
+
+## Defining an Inferred Type
+
+The `typeof` function can be used in a type definition to define an inferred type.
+
+```lua
+type Car = typeof {
+	Speed = 0,
+	Wheels = 4
+} --> Car: {Speed: number, Wheels: number}
+```
+
+One way to use `typeof`, is to define a metatable type using `setmetatable` inside the `typeof` function.
+
+```lua
+type Vector = typeof(setmetatable({}::{
+	x: number,
+	y: number
+}, {}::{
+	__add: (Vector, Vector|number) -> Vector
+}))
+
+-- Vector + Vector would return a Vector type

```

Comment:
I'm super unclear about what's going on here. Can we create a couple of these, add them together, and show the output...?